### ISSUE TITLE
tf.contrib.boosted_trees cannot be used in 1.4.0rc1

### ISSUE THREAD

**Issue by zhedongzheng** [tensorflow/tensorflow#14087]
TF Version: 1.4.0rc1
Py Version: 2.7
OS: Mac OS

```python
import tensorflow as tf
from tensorflow.contrib.boosted_trees.estimator_batch.estimator import GradientBoostedDecisionTreeClassifier
from tensorflow.contrib.boosted_trees.proto import learner_pb2

learner_config = learner_pb2.LearnerConfig()

learner_config.learning_rate_tuner.fixed.learning_rate = 0.1
learner_config.num_classes = 10
learner_config.regularization.l1 = 0.0
learner_config.regularization.l2 = 1.0 / 1000
learner_config.constraints.max_tree_depth = 4
learner_config.growing_mode = learner_pb2.LearnerConfig.LAYER_BY_LAYER
learner_config.multi_class_strategy = (learner_pb2.LearnerConfig.DIAGONAL_HESSIAN)

estimator = GradientBoostedDecisionTreeClassifier(
    learner_config = learner_config,
    n_classes = 10,
    examples_per_layer = 1000,
    num_trees = 10,
    center_bias = False)

(X_train, y_train), (X_test, y_test) = tf.keras.datasets.mnist.load_data()
X_train = (X_train / 255.).reshape(-1, 28*28).astype(np.float32)
y_train = y_train.astype(np.int32)

estimator.fit(input_fn=tf.estimator.inputs.numpy_input_fn(
    x={'features':X_train}, y=y_train, batch_size=128, num_epochs=1, shuffle=True))
```

A minimal example to train on mnist dataset throws error:
```
Traceback (most recent call last):
  File "boost_tree.py", line 29, in <module>
    x={'features':X_train}, y=y_train, batch_size=128, num_epochs=1, shuffle=True))
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/python/util/deprecation.py", line 316, in new_func
    return func(*args, **kwargs)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 480, in fit
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 986, in _train_model
    model_fn_ops = self._get_train_ops(features, labels)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1202, in _get_train_ops
    return self._call_model_fn(features, labels, model_fn_lib.ModeKeys.TRAIN)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/estimator.py", line 1166, in _call_model_fn
    model_fn_results = self._model_fn(features, labels, **kwargs)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/boosted_trees/estimator_batch/model.py", line 116, in model_builder
    logits=logits)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 1064, in create_model_fn_ops
    enable_centered_bias=self._enable_centered_bias)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 648, in _create_model_fn_ops
    batch_size, loss_fn, weight_tensor)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/estimators/head.py", line 1923, in _train_op
    train_op = train_op_fn(loss)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/boosted_trees/estimator_batch/model.py", line 105, in _train_op_fn
    update_op = gbdt_model.train(loss, predictions_dict, labels)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/boosted_trees/python/training/functions/gbdt_batch.py", line 543, in train
    hessian_list = self._diagonal_hessian(gradients, predictions)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/contrib/boosted_trees/python/training/functions/gbdt_batch.py", line 845, in _diagonal_hessian
    aggregation_method=None)
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/python/ops/gradients_impl.py", line 581, in gradients
    grad_scope, op, func_call, lambda: grad_fn(op, *out_grads))
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/python/ops/gradients_impl.py", line 353, in _MaybeCompile
    return grad_fn()  # Exit early
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/python/ops/gradients_impl.py", line 581, in <lambda>
    grad_scope, op, func_call, lambda: grad_fn(op, *out_grads))
  File "/Users/zhedongzheng/pyenv/py2.7/tf_1.4/tf_1.4/lib/python2.7/site-packages/tensorflow/python/ops/array_grad.py", line 352, in _PreventGradientGrad
    "Gradient explicitly disabled. Reason: %s" % op.get_attr("message"))
LookupError: Gradient explicitly disabled. Reason: Currently there is no way to take the second derivative of sparse_softmax_cross_entropy_with_logits due to the fused implementation's interaction with tf.gradients()
```

**Comment by asimshankar** [comment]
@tkoeppe : Mind taking a look?

**Comment by tkoeppe** [comment]
@asimshankar: Do you mean because of my BUILD cleanups? All the tests passed, but maybe. @zhedongzheng, would you mind testing with the relevant changes from 966016b7f2382658e7c84baae0596d35f0a49bae and 6ef542822b8760de4beeb4ad2602fb863888d47f reverted?

**Comment by zhedongzheng** [comment]
@tkoeppe hi, i want to help, but i don't really know how to go back to previous commits in the codebase.

**Comment by tkoeppe** [comment]
@zhedongzheng: Oh, I just meant to edit the changes back in manually. No need for fancy git gymnastics.

**Comment by zhedongzheng** [comment]
@tkoeppe The code does not work in TF 1.3 either, so i don't think it is a BUILD problem, it is an implementation problem i suppose.

**Comment by zhedongzheng** [comment]
@ispirmustafa any insight about this? or do you know who should be called?

**Comment by tkoeppe** [comment]
@asimshankar: I'm not otherwise terribly familiar with TF, sorry! Back to you?

@zhedongzheng: Thanks for investigating!

**Comment by tkoeppe** [comment]
One of my colleagues suggested that second-order gradients are not supported well across all of TF, which is what the error message is trying to say. Does that help? Has this code ever worked in the past?

**Comment by sshrdp** [comment]
The issue has been fixed in 34a4b21f8f9dea64d3e99a97f639396f2d5556d3 by switching to a custom loss function that had support for second order gradient within the GBDTClassifier estimator.  You can wait for the release of that change or see boosted_trees/examples/mnist.py [before the above commit](https://github.com/tensorflow/tensorflow/blob/c41dbc3c1832bc6c3662d4d942d095baa1fb49c9/tensorflow/contrib/boosted_trees/examples/mnist.py#L92) which uses the custom loss function.
